### PR TITLE
Use container-based infrastructure

### PR DESCRIPTION
**Pull request by tigerw**
The problem here is that, since Travis container doesn't yet support [update-alternatives](https://github.com/travis-ci/travis-ci/issues/3668), `export CXX` will set the compiler to GCC 5 even if it's a Clang build.

But it works, amiright?

### CODE REVIEW

**Comment by tigerw** [comment]
Travis fails due to Google-test.

**Comment by worktycho** [comment]
How about only changing the CC and CXX cars if we are in a gcc build?

**Comment by tigerw** [comment]
Good idea. Any idea why Google-test is throwing an exception, but only on one build?

**Comment by tigerw** [comment]
Google test still randomly excepting for reasons unknown. Would this be mergeable?

**Comment by madmaxoft** [comment]
I think the google test has something to do with LibEvent deinitialization being done at a wrong moment, but I have no idea on how to ascertain that, let alone fix it.

**Comment by tigerw** [comment]
So merge?

**Comment by madmaxoft** [comment]
Rebase onto current master (and get rid of your non-working google fix) to see if it actually compiles well.

**Comment by tigerw** [comment]
Google was returning a 302 without the www prefix so that commit is still valid. However, since I'm in 'murica I can't do anything fancy :/

**Comment by tigerw** [comment]
If you examine the logs, the failure is always during the test - builds start and run fine.

**Comment by madmaxoft** [comment]
I've rebased onto current master and the builds fail.

**Comment by madmaxoft** [comment]
Seems like `clang-3.7` is not available. Perhaps in another repository? Or use another version?

**Comment by tigerw** [comment]
Still not sure which repository it's in though.

**Comment by worktycho** [comment]
clang 3.7 doesn't exist yet. Testing against 3.4 is important for warnings. Is it possible to test against 3.4 and 3.6?

**Comment by tigerw** [comment]
I was sure it managed to install 3.7 before, how curious.

Testing against both sounds like it would require further extensinos to the condition in CIBuild.sh - I guess a test for Clang and Clang version. If 3.7 has less warnings than 3.4 (amazing), how about I remain on 3.4?

**Comment by worktycho** [comment]
3.7 is currently the trunk provisional version, so maybe you installed trunk?

**Comment by tigerw** [comment]
yes!

**Comment by tigerw** [comment]
I mean, yes it compiled! as opposed to yes I installed trunk!
I'm not sure what I installed.

**Comment by madmaxoft** [comment]
I thought we wanted to use clang 3.6? Or am I just too old and forgetful?

**Comment by madmaxoft** [comment]
Also, any chance of getting rid of that `gcc --version` in favor of `gcc-4.8 --version` in the gcc builds?

File: src/Bindings/LuaState.cpp [hunk 1]
```diff
@@ -868,7 +868,7 @@ cLuaState::cStackValue cLuaState::WalkToValue(const AString & a_Name)
 		// Remove the previous value from the stack (keep only the new one):
 		lua_remove(m_LuaState, -2);
 	}  // for elem - path[]
-	return std::move(cStackValue(*this));
+	return cStackValue(*this);
```
Comment (madmaxoft) [inline src/Bindings/LuaState.cpp]: I'm not too confident about this change.

File: CIbuild.sh [hunk 2]
```diff
@@ -6,6 +6,9 @@ export MCSERVER_BUILD_SERIES_NAME="Travis $CC $TRAVIS_MCSERVER_BUILD_TYPE"
 export MCSERVER_BUILD_ID=$TRAVIS_JOB_NUMBER
 export MCSERVER_BUILD_DATETIME=`date`
 
+if [ "$CXX" == "g++" ]; then
+	export CXX="/usr/bin/g++-4.8"
+fi
```
Comment (madmaxoft) [inline CIbuild.sh]: This could use a comment saying why you're overriding the CXX variable. Also, shouldn't CC be overridden as well?

File: src/Bindings/LuaState.cpp (hunk 1 above)
Comment (tigerw) [inline src/Bindings/LuaState.cpp]: Clang 3.7 said to remove std::move else it wouldn't compile.

File: CIbuild.sh (hunk 2 above)
Comment (tigerw) [inline CIbuild.sh]: Okay.
It works without CC or an upgrade to gcc-4.8, why bother :P

**Comment by tigerw** [comment]
Tycho suggested testing with both clang 3.4 and 3.6, but I'm having difficulty getting a clang upgrade to work, so I just left it at 3.4 + gcc 4.8.

However, it would be nice to do something with clang since gcc 5 libraries don't appear to work with older clang versions.

File: src/Bindings/LuaState.cpp (hunk 1 above)
Comment (tigerw) [inline src/Bindings/LuaState.cpp]: What's the difference?

**Comment by tigerw** [comment]
I'm not precisely sure how to change the redirection of `gcc` to `gcc-4.8` without (sudo) update-alternatives. Any ideas?

File: src/Bindings/LuaState.cpp (hunk 1 above)
Comment (madmaxoft) [inline src/Bindings/LuaState.cpp]: I think the original code used move-construction, while the new code could be using copy-construction (which isn't exactly supported by the `cStackValue` class)

Comment (tigerw) [inline src/Bindings/LuaState.cpp]: The compiler was complaining though. See first comment for error: https://news.ycombinator.com/item?id=9915300

```
    foo.cpp:44:25: error: moving a temporary object prevents copy elision [-Werror,-Wpessimizing-move]
        auto result = std::move(worker.run());

    foo.cpp:44:25: note: remove std::move call here
        auto result = std::move(worker.run());
```

**Comment by SphinxC0re** [comment]
What about *Use container-based infrastructure with Travis*?

**Comment by tigerw** [comment]
On [this](https://github.com/travis-ci/travis-ci/issues/3668) issue it is reported that they're looking into it.

**Comment by tigerw** [comment]
oops, I made if "for" instead of "with". Is this okay?

**Comment by tigerw** [comment]
I guess it is! meeeeeeeeerge --->

File: SetFlags.cmake [hunk 3]
```diff
@@ -266,11 +266,18 @@ macro(set_exe_flags)
...
+				check_cxx_compiler_flag(-Wno-documentation-unknown-command HAS_NO_DOCUMENTATION_UNKNOWN)
 				if (HAS_NO_RESERVED_ID_MACRO)
 					# Use this flag to ignore error for a reserved macro problem in sqlite 3
 					add_flags_cxx("-Wno-reserved-id-macro")
 				endif()
+				if (HAS_NO_DOCUMENTATION_UNKNOWN)
+					# Ignore another problem in sqlite
+					add_flags_cxx("-Wno-documentation-unknown-command")
+				endif()
+				add_flags_cxx("-Wno-format-pedantic")
```
Comment (worktycho) [inline SetFlags.cmake]: I didn't even notice this. Why did you add it?

Comment (tigerw) [inline SetFlags.cmake]: Ref: https://github.com/paul99/v8m-rb/issues/128

3.7 builds failed without it. Clang complains tye %p specifier expects `void *` but that rarely happened.

Comment (worktycho) [inline SetFlags.cmake]: Are we supporting pre-release compilers now?

Comment (tigerw) [inline SetFlags.cmake]: yes :)

Comment (worktycho) [inline SetFlags.cmake]: Then can we define what we mean by c++? because technically clang is right that passing anything other than a `void *` to %p is UB.

Comment (tigerw) [inline SetFlags.cmake]: C++ is a programming language based off C, no?

If UggBoots are the case, then it is of course necessary to fix those instances. I shall inform #2441 to remove the flag after all.

File: .travis.yml [hunk 4]
```diff
@@ -1,20 +1,17 @@
...
-
-  # lua, needed for style checking and possibly later on for bindings generation
-  - sudo apt-get install -qq lua5.1
+addons:
+  apt:
+    sources:
+    - ubuntu-toolchain-r-test
+    packages:
+    - g++-4.8
+    - lua5.1
```
Comment (worktycho) [inline .travis.yml]: Another thing I've just noticed. Why are we installing lua, when we don't use the system lua in the build?

Comment (madmaxoft) [inline .travis.yml]: We use Lua sources and headers, don't we?

Comment (worktycho) [inline .travis.yml]: The ones we bundle. This asks for their to be a system lua, which we then don't use.

Comment (tigerw) [inline .travis.yml]: All I did was convert the previous `apt-get install lua5.1` to this.

Comment (tigerw) [inline .travis.yml]: Apparently I did remove a comment though: https://github.com/cuberite/cuberite/commit/e8c8aca7791a598244aa9dcc9a0992d2c249652d

Comment (worktycho) [inline .travis.yml]: Ah, its used for the style check in the CI script.
